### PR TITLE
[PR] Support node micro layout with further layout algorithms

### PR DESCRIPTION
**Pull request by uruuru**
Algorithms
* force
* stress
* mrtree
* radial
* rectpacking 

now call the code that computes node label positions, node dimensions, port positions, and port label positions. 

As I'm not absolutely certain that no negative side-effects occur for these layoutes when running the placement code beforehand, I introduced the `omitNodeMicroLayout` option to have a possibility to silence the code :).

### CODE REVIEW

File: plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/NodeMicroLayout.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Utility class to execute what can be considered as "node micro layout". That is, automatically computing node
+ * dimensions if requested by the user, positioning ports, positioning node and port labels, etc.
+ * 
+ * Currently, simply all possible micro layouts are executed. It may make sense to perform only a subset, e.g. only
+ * place node labels. This may be implemented in the future.
+ * 
+ * Furthermore, all of {@link NodeDimensionCalculation}'s methods should be moved to this utility class at some point as
+ * this class's naming makes it a bit clearer what actually happens.
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/NodeMicroLayout.java:25]: Oh yes, please!

File: plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/NodeMicroLayout.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        return forGraph(ElkGraphAdapters.adapt(elkGraph));
+    }
+
+    /**
+     * @return a new micro layout instance for the passed graph. Use {@link NodeMicroLayout#execute()} to finally
+     *         perform the layout process.
+     */
+    public static NodeMicroLayout forGraph(final GraphAdapter<?> adapter) {
+        NodeMicroLayout builder = new NodeMicroLayout(adapter);
+        return builder;
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/NodeMicroLayout.java:49]: Why not shorten this to a single line?

File: plugins/org.eclipse.elk.alg.force/src/org/eclipse/elk/alg/force/Force.melk [hunk 3]
```diff
@@ -46,6 +46,12 @@ algorithm force(ForceLayoutProvider) {
     supports org.eclipse.elk.padding = new ElkPadding(50)
     supports org.eclipse.elk.interactive
     supports org.eclipse.elk.portConstraints
+    // Common node micro layout
+    supports org.eclipse.elk.omitNodeMicroLayout
+    supports org.eclipse.elk.nodeSize.options
+    supports org.eclipse.elk.nodeSize.constraints
+    supports org.eclipse.elk.nodeLabels.placement
+    supports org.eclipse.elk.portLabels.placement
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.force/src/org/eclipse/elk/alg/force/Force.melk:54]: Are these the only options the node placement code uses?

Edit: For example, the stress layouter also support minimum node size?

File: plugins/org.eclipse.elk.alg.force/src/org/eclipse/elk/alg/force/ForceLayoutProvider.java [hunk 4]
```diff
@@ -36,15 +35,21 @@
...
+            NodeMicroLayout.forGraph(elkGraph)
+                           .execute();
+        }
+        
         // transform the input graph
         IGraphImporter<ElkNode> graphImporter = new ElkGraphImporter();
-        FGraph fgraph = graphImporter.importGraph(kgraph);
+        FGraph fgraph = graphImporter.importGraph(elkGraph);
 
         // set special properties for the layered graph
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.force/src/org/eclipse/elk/alg/force/ForceLayoutProvider.java:51]: "layered graph" 😄

File: plugins/org.eclipse.elk.alg.force/src/org/eclipse/elk/alg/force/stress/StressLayoutProvider.java [hunk 5]
```diff
@@ -37,9 +36,17 @@
     public void layout(final ElkNode layoutGraph, final IElkProgressMonitor progressMonitor) {
         progressMonitor.begin("ELK Stress", 1);
 
+
         // calculate initial coordinates
         if (!layoutGraph.getProperty(StressOptions.INTERACTIVE)) {
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.force/src/org/eclipse/elk/alg/force/stress/StressLayoutProvider.java:41]: Not negating the condition and swapping the two branches would make this a little more readable. I always have to think to figure out when the `else` branch is executed.

File: plugins/org.eclipse.elk.alg.radial/src/org/eclipse/elk/alg/radial/RadialLayoutProvider.java [hunk 6]
```diff
@@ -53,12 +60,6 @@ public void layout(final ElkNode layoutGraph, final IElkProgressMonitor progress
         }
         layoutGraph.setProperty(RadialOptions.RADIUS, layoutRadius);
         
-        //add support for the NodeDimensionCaluclation
-        ElkGraphAdapter adapter = ElkGraphAdapters.adapt(layoutGraph);
-        layoutGraph.getWidth();
-        NodeDimensionCalculation.calculateLabelAndNodeSizes(adapter);   
-        NodeDimensionCalculation.calculateNodeMargins(adapter);
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.radial/src/org/eclipse/elk/alg/radial/RadialLayoutProvider.java:60]: The code replacing this changes the layouter's behavior in that it also sorts the port lists. Is that intended?

File: plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/Core.melk [hunk 7]
```diff
@@ -285,6 +285,18 @@ advanced option interactiveLayout: boolean {
     targets parents
 }
 
+advanced option omitNodeMicroLayout: boolean {
+    label "Omit Node Micro Layout"
```
Comment (le-cds) [inline plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/Core.melk:289]: This is a "negated" layout option in that setting it to `true` actually _disables_ something. I suggest doing it the other way around: calling it `runNodeMicroLayout` and changing its default to `true`.

File: plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/Core.melk [hunk 8]
```diff
@@ -285,6 +285,18 @@ advanced option interactiveLayout: boolean {
...
+    description
+        "Node micro layout comprises the computation of node dimensions (if requested), the placement of ports
```
Comment (le-cds) [inline plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/Core.melk:291]: "if requested" refers to node size constraints, right? I think it would be good to reference them here.

**Review by le-cds — CHANGES_REQUESTED** [review]

**Review by soerendomroes — COMMENTED** [review]
Despite already mentioned things this looks good to me.

File: plugins/org.eclipse.elk.alg.force/src/org/eclipse/elk/alg/force/Force.melk (hunk 3 above)
Comment (uruuru) [reply plugins/org.eclipse.elk.alg.force/src/org/eclipse/elk/alg/force/Force.melk:54]: I must have missed the minimum size here - apart from that it's a best guess.

File: plugins/org.eclipse.elk.alg.radial/src/org/eclipse/elk/alg/radial/RadialLayoutProvider.java (hunk 6 above)
Comment (uruuru) [reply plugins/org.eclipse.elk.alg.radial/src/org/eclipse/elk/alg/radial/RadialLayoutProvider.java:60]: Yes. I don't see any reason not to execute the port list sorting. Without giving it much thought though.

File: plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/Core.melk (hunk 7 above)
Comment (uruuru) [reply plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/Core.melk:289]: I felt that - since running micro layout is the default - a user would look for a layout option to _omit_ that behavior.

I'm not sure what is more intuitive to most people though. Yours maybe reduced cognitive load :D

**Comment by uruuru** [comment]
Shall we include this in 0.7.1?

**Comment by le-cds** [comment]
Personally, I think this is a new feature and should be targeted for 0.8.0.

**Comment by uruuru** [comment]
Agreed. Corresponds to my tendency.
